courses/Ada_For_The_C_Embedded_Developer/chapters/02_Perspective.rst

```diff
@@ -0,0 +1,802 @@
+      int sum = 0;
+
+      for(i = 0; i < length; ++i) {
+         sum += list[i];
+      }
+      return (sum / length);
+   }
+
+.. code-block:: c
+
+   // The modern C way
+   int average(int* list, int length)
+   {
+      int sum = 0;
+
+      for(int i = 0; i < length; ++i) {
+         sum += list[i];
+      }
+
+      return (sum / length);
+   }
+
+For the fun of it, let's also see the Ada way to do this:
+
+.. code-block:: ada
+
+   type Int_Array is array (Natural range <>) of Integer;
+
+   function Average (List : Int_Array) return Integer
+   is
+      Sum : Integer;
+   begin
+
+      for I in List'Range loop
+         Sum := Sum + List (I);
+      end loop;
+
+      return (Sum / List'Length);
+   end Average;
+
+We will explore more about the syntax of loops in Ada in a future section of this book; but for now, notice that the :ada:`I` variable used as the loop index is not declared in the declarative section!
+
+.. admonition:: Declaration Flippy Floppy
+
+   Something peculiar that you may have noticed about declarations in Ada is that they are backwards from the way C does declarations. The C language expects the type followed by the variable name. Ada expects the variable name followed by a semicolon and then the type.
+
+The next block in the Ada example is between the :ada:`begin` and :ada:`end` keywords. This is where your statements will live. You can create new scopes by using the :ada:`declare` keyword:
+
+.. code:: ada
+
+   with Ada.Text_IO;
+
+   procedure Main
+   is
+      --  variable declaration
+      A, B : Integer := 0;
+      C    : Integer := 100;
+      D    : Integer;
+   begin
+      --  Ada does not have a shortcut format for increment like in C
+      A := A + 1;
+
+      --  regular addition
+      D := A + B + C;
+
+      --  printing the result
+      Ada.Text_IO.Put_Line ("D =" & D'Img);
+
+      declare
+         E : Integer := D * 100;
+      begin
+         --  printing the result
+         Ada.Text_IO.Put_Line ("E =" & E'Img);
+      end;
+
+   end Main;
+
+Notice that we declared a new variable :ada:`E` whose scope only exists in our newly defined block. The equivalent C code is:
+
+.. code:: c
+
+   !main.c
+   #include <stdio.h>
+
+   int main()
+   {
+      // variable declarations
+      int a = 0, b = 0, c = 100, d;
+
+      // c shorthand for increment
+      a++;
+
+      // regular addition
+      d = a + b + c;
+
+      // printing the result
+      printf("d = %d\n", d);
+
+      {
+         int e = d * 100;
+         printf("e = %d\n", e);
+      }
+
+      return 0;
+   }
+
+.. admonition:: The shortcuts of incrementing and decrementing
+
+   You may have noticed that Ada does not have something similar to the :c:`a++` or :c:`a--` operators. Instead you must use the full assignment :ada:`A := A + 1` or :ada:`A := A - 1`.
+
+**Fun Fact** about the C language assignment operator :c:`=`: Did you know that an assignment in C can be used in an expression? Let's look at an example:
```

Comment:
I assume most C developers know this, so I'd rephrase this as a statement.

Reply:
I disagree that most C developers know this. There are CWE's about this:
https://cwe.mitre.org/data/definitions/481.html

and MISRA has an explicit rule about this. You're welcome to rephrase it into a statement though.